lib/mongo/crypt/encryption_io.rb

```diff
@@ -95,29 +95,58 @@ def mark_command(cmd)
       #   the endpoint at which to establish a TLS connection and the message
       #   to send on that connection.
       def feed_kms(kms_context)
-        endpoint = kms_context.endpoint
-        message = kms_context.message
+        with_ssl_socket(kms_context.endpoint) do |ssl_socket|
+          ssl_socket.puts(kms_context.message)
 
+          while kms_context.bytes_needed > 0 do
+            bytes = ssl_socket.sysread(kms_context.bytes_needed)
+            kms_context.feed(bytes)
+          end
+        end
+      end
+
+      private
+
+      # Provide an SSL socket to be used for KMS calls in a block API
+      #
+      # @param [ String ] endpoint The URI at which to connect the SSL socket
+      # @param [ Proc ] block The block to execute
+      #
+      # @raise [ Mongo::Error::KmsError ] If the socket times out or raises
+      #   an exception
+      #
+      # @note The socket is always closed when the provided block has finished
+      #   executing
+      def with_ssl_socket(endpoint)
         # There is no specific timeout written in the spec. See SPEC-1394
         # for a discussion and updates on what this timeout should be.
         socket_timeout = 10
+        default_port = 443
 
-        host, port = endpoint.split(':')
-        port ||= 443
+        unless endpoint.start_with?('https://') || endpoint.start_with?('//')
+          endpoint = "//#{endpoint}"
+        end
```

Comment:
My goal here is to make sure that URIs are correctly parsed in as many situations as possible. By default, LMC returns the KMS endpoint in the format `kms.us-east-1.amazonaws.com`, which confuses Ruby's URI parser because there is no scheme. Is there a better or more reliable way to handle this case?

Reply:
https://github.com/mongodb/libmongocrypt/blob/master/integrating.md#state-mongocrypt_ctx_need_kms says:

> Create/reuse a TLS socket connected to the endpoint indicated by mongocrypt_kms_ctx_endpoint. The endpoint string is a host name with optional port number separated by a colon

So, this is actually supposed to be a URI, but just the host name, right?

Reply:
Okay. In that case, do you think it's fair for me to just split the string at the colon and then raise an exception if the result doesn't have two parts? I can also specify in the documentation that any user-specified endpoint should take this same format.

Reply:
Yes per our discussion just now.